src/dsp/format_converter.h

```diff
@@ -0,0 +1,144 @@
+/* -*- c++ -*- */
+/*
+ * Gqrx SDR: Software defined radio receiver powered by GNU Radio and Qt
+ *           https://gqrx.dk/
+ *
+ * Copyright 2021 vladisslav2011@gmail.com.
+ *
+ * Gqrx is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * Gqrx is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Gqrx; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+#ifndef INCLUDED_FORMAT_CONVERTER_H
+#define INCLUDED_FORMAT_CONVERTER_H
+
+#include <gnuradio/blocks/api.h>
+#include <gnuradio/sync_interpolator.h>
+#include <volk/volk.h>
+#include <iostream>
+
+namespace dispatcher
+{
+    template <typename> struct tag{};
+}
+
+    /*!
+     * \brief Convert stream of one format to a stream of another format.
```

Comment:
Can't you just use GNU Radio's built-in type conversion blocks for this?

Reply:
I've did it before and switched to my own implementation later because:
1. GNU Radio's built-in type conversion blocks are not using libvolk (for some unknown reason), so performance is not great.
2. My GNU Radio installation (distribution supplied, yes, I'm lazy to compile it myself) lacks interleaved_char_to_complex and complex_to_interleaved_char blocks.
3. Using GNU Radio blocks requires to chain a multiply_const block for scaling. Libvolk functions do both conversion and scaling at one call.

Reply:
As with the File Sink, I don't think Grqx is the right place to do this. I don't want the added maintenance cost of maintaining a parallel implementation.

> GNU Radio's built-in type conversion blocks are not using libvolk

I'm sure GNU Radio would appreciate a patch to fix this.

> My GNU Radio installation lacks interleaved_char_to_complex and complex_to_interleaved_char blocks.

What version are you using, and what version were these blocks added in? If needed, we could use a cascade of blocks in Gqrx to support older versions of GNU Radio.

> Using GNU Radio blocks requires to chain a multiply_const block for scaling.

Some of the type conversion blocks have built-in scaling. If others lack it, then I'd suggest contributing a patch to GNU Radio.

Reply:
> I'm sure GNU Radio would appreciate a patch to fix this.
This is already done in GNU Radio 3.9.1 but setting minimum required GNU Radio version to 3.9.1 would break compilation on many major distributions.

> If needed, we could use a cascade of blocks in Gqrx to support older versions of GNU Radio.
I've written before, that I've did it this way first. But then I have found, that complex_to_interleaved_char block is not shipped by my base distribution (Ubuntu 18.04LTS). So I have to either build recent GNU Radio from source on all machines, where I use gqrx (different distributions, different architectures), or pull the block from GNU Radio source code, or replace all conversion blocks with small template class, that may be easily extended to support more formats later (it is possible as we do not have to support python binding). 
Cascade blocks are bad from performance point of view: more unneeded buffers, more threads.
Sometimes all major distributions will adopt GNU Radio 3.9.1 and we'll be able to get rid of this small header.
As an option, I can convert added blocks to implement modern GNU Radio interface (inherit or just typedef would be enough?) and put the implementation into "compat" folder, that would not get compiled/included if we have found GNU Radio 3.9.1. This would reduce maintenance time later. But this would add more #ifdefs and grow CMakelists.txt...

Reply:
My suggestion would be to have two different versions of the flow graph, depending on the GNU Radio version. Gqrx already does that in a few places, for example:

https://github.com/gqrx-sdr/gqrx/blob/75d108f7e9b0bb69bb1847173d957b8e41a0cc8f/src/dsp/rx_rds.cpp#L85-L103

I would suggest to use a cascade of blocks for type conversion in older versions, and the new IChar blocks (starting in 3.8?).

As a side note, supporting GNU Radio 3.7 is already becoming a burden, so I'd like to drop support for it soon. Users on old distributions can always use the AppImage releases.